agent/xds/server.go

```diff
@@ -176,8 +193,53 @@ func (s *Server) process(stream ADSStream, reqCh <-chan *envoy.DiscoveryRequest)
 		},
 	}
 
+	var timer *time.Timer
+	var authTimerChan <-chan time.Time
+	extendAuthTimer := func(timerReceived bool) {
+		if timer == nil {
+			return
+		}
+		if !timerReceived {
+			if !timer.Stop() {
+				<-timer.C
```

Comment:
I'd leave a comment pointing out that per-docs we need to drain the channel if the timer is already stopped.

Having looked it up I'm still not sure why this is correct - shouldn't we only drain the timer the first time we stop (i.e. time.Stop() returns true not false)? I guess it's right but it's certainly not intuitive to me and could be clearer with comments IMO! I think this is also assuming that no other code stops or drains the timer (reasonable but not stated) otherwise this could block forever.

Reply:
Twice while writing this I ended up blocking the RPC forever by getting this `Timer.Reset` dance wrong. I'll add some comments.

Reply:
I actually just checked `hashicorp/raft` to see how extendable timers were used in heartbeat logic. Every heartbeat a new `time.After` recv chan is returned. I think I'm just going to use that instead as it doesn't have nearly as many sharp edges and the diff is cleaner. If it's good enough for raft it should be good enough here.